playbooks/configure-apt-sources.yml

```diff
@@ -119,48 +119,129 @@
+          changed_when: false
+          delegate_to: "{{ physical_host | default(omit) }}"
+          when:
+            - host_ubuntu_repo is not defined
+
+        - name: Set host_ubuntu_repo fact
+          set_fact:
+            host_ubuntu_repo: "{{ _ubuntu_repo.stdout_lines[0] }}"
+          when:
+            - host_ubuntu_repo is not defined
+            - _ubuntu_repo.stdout_lines is defined
 
-    - name: Determine the existing Ubuntu repo configuration
-      shell: 'sed "s/^[ \t]*//" /etc/apt/sources.list | grep -oP "^deb \K(\[?.*\]?.*ubuntu\S*\/?)(?= {{ ansible_distribution_release }} main)"'
-      register: _ubuntu_repo
-      changed_when: false
-      delegate_to: "{{ physical_host | default(omit) }}"
+        - name: Replace the apt sources file with our content
+          copy:
+            content: |
+              # Base repository
+              deb {{ host_ubuntu_repo }} {{ ansible_distribution_release }} main universe
+            dest: "/etc/apt/sources.list"
+            backup: yes
+          register: apt_sources_configure
+
+        - name: Create the rpco apt sources file
+          copy:
+            content: |
+              # RPC-OpenStack repository
+              deb {{ rpco_mirror_base_url }}/apt-mirror/integrated/ {{ rpc_release }}-{{ ansible_distribution_release }} main
+            dest: "/etc/apt/sources.list.d/rpco.list"
+            backup: yes
+          register: apt_sources_configure_rpco
+
+        - name: Add rpco keys
+          apt_key:
+            url: "{{ rpco_mirror_base_url }}/apt-mirror/rcbops-release-signing-key.asc"
+            state: "present"
+          register: add_keys
+          until: add_keys | success
+          retries: 5
+          delay: 2
+
+        - name: Remove extra sources
+          lineinfile:
+            dest: /etc/apt/sources.list
+            state: absent
+            regexp: "{{ item }}"
+          with_items:
+            - "^deb-src"
+            - "-backports"
+            - "-security"
+            - "-updates"
       when:
-        - host_ubuntu_repo is not defined
         - apt_artifact_found | bool
         - apt_artifact_enabled | bool
 
-    - name: Set host_ubuntu_repo fact
-      set_fact:
-        host_ubuntu_repo: "{{ _ubuntu_repo.stdout_lines[0] }}"
+    - name: Revert apt artifacts
+      block:
+        - name: Remove artifact group vars (all)
+          file:
+            dest: "/etc/openstack_deploy/group_vars/all/{{ item | basename }}"
+            state: absent
+          with_items: "{{ artifact_apt_group_files }}"
+          run_once: true
+          delegate_to: localhost
+
+        - name: Remove artifact user vars (all)
+          file:
+            dest: "/etc/openstack_deploy/group_vars/all/{{ item | basename }}"
+            state: absent
+          with_items: "{{ artifact_apt_user_files }}"
+          run_once: true
+          delegate_to: localhost
+
+        - name: Check for original sources file
```

Comment:
In the current environment, the backup is always created if it's not there, negating the need to have this task and always ensuring that the original is backed up. Why not leave that in place and do away with the extra tasks?

Reply:
done.